.github/workflows/pdm-lock-automation.yaml

```diff
@@ -0,0 +1,95 @@
+name: Automate pdm.lock
+
+on: push
+
+jobs:
+  check_pyproject_changes:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        directory: [backend, prompt-service, root]
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Check file changes
+        id: check
+        run: |
+          if [ "${{ matrix.directory }}" = "root" ]; then
+            file_path="pyproject.toml"
+          else
+            file_path="${{ matrix.directory }}/pyproject.toml"
+          fi
+
+          echo "checking changes for $file_path"
+          echo "changed=false" >> $GITHUB_ENV
+          if git diff --quiet main -- $file_path; then
+            echo "No changes detected in $file_path against main branch."
+          else
+            echo "Changes detected in $file_path against main branch."
+            echo "changed=true" >> $GITHUB_ENV
+          fi
+
+      - name: Set output
+        id: set_output
+        run: echo "::set-output name=changed::${{ env.changed }}"
+        if: always()
+
+      - name: Set up Python
+        if: steps.set_output.outputs.changed == 'true'
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.9'
+
+      - name: Set up Go
+        if: steps.set_output.outputs.changed == 'true'
+        uses: actions/setup-go@v4
+        with:
+          go-version: '1.22.5'
+
+      - name: Install PDM
+        if: steps.set_output.outputs.changed == 'true'
+        run: python -m pip install pdm
```

Comment:
@kirtimanmishrazipstack might be a good idea to lock this to a version that our Dockerfiles also expect / work with to avoid any surprises